vendor/noise_handshake/src/handshake.rs

```diff
@@ -0,0 +1,217 @@
+                },
+                noise: Some(self),
+                io: Some(io),
+            },
+            _ => panic!(
+                "Noise Session in incorrect state, you should init before starting handshake"
+            ),
+        }
+    }
+}
+
+#[allow(missing_debug_implementations)]
+pub struct NoiseHandshake<IO: AsyncRead + AsyncWrite> {
+    next: Step,
+    noise: Option<Session>,
+    io: Option<IO>,
+}
+
+impl<IO: AsyncRead + AsyncWrite> NoiseHandshake<IO> {
+    fn wrap_up(&mut self) -> (Session, IO) {
+        let noise = self.noise.take().expect("We know it's a Some");
+        let noise = noise
+            .into_transport_mode()
+            .expect("Cannot go into transport mode despite handshake being finished");
+        let io = self.io.take().expect("We know it's a Some");
+        (noise, io)
+    }
+}
+
+impl<IO: AsyncRead + AsyncWrite> Future for NoiseHandshake<IO> {
+    type Item = (Session, IO);
+    type Error = std::io::Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        match self {
+            Self {
+                noise: Some(ref mut noise),
+                io: Some(_),
+                next:
+                    Step::Write {
+                        ref mut buffer,
+                        ref mut len,
+                        ..
+                    },
+            } if *len == 0 => {
+                *len = noise
+                    .write_message(&[], buffer)
+                    .expect("Cannot encode the message");
+                self.poll()
+            }
+            Self {
+                noise: Some(ref mut noise),
+                io: Some(ref mut io),
+                next:
+                    Step::Write {
+                        buffer,
+                        ref mut written_bytes,
+                        len,
+                    },
+                ..
+            } => {
+                while written_bytes < len {
+                    *written_bytes = try_ready!(io.poll_write(&buffer[*written_bytes..*len]));
+                }
+                if noise.is_handshake_finished() {
+                    Ok(Async::Ready(self.wrap_up()))
+                } else {
+                    self.next = Step::read();
+                    self.poll()
+                }
+            }
+            Self {
+                noise: Some(ref mut noise),
+                io: Some(ref mut io),
+                next:
+                    Step::Read {
+                        mut enc_buffer,
+                        ref mut len,
+                    },
+            } => {
+                let mut dec_buffer = [0u8; NOISE_MAX_SIZE];
+
+                *len += try_ready!(io.poll_read(&mut enc_buffer[*len..]));
+                if *len > 0 {
+                    match noise.read_message(&enc_buffer[..*len], &mut dec_buffer) {
+                        Ok(_) => self.next = Step::write(),
+                        Err(e) => debug!("Error decoding message: {:?}", e),
```

Comment:
Knowing the size of what we are expecting would remove this weird Err debug thing. If we know what we are looking for we can just return a proper error.